### ISSUE TITLE
Gradle plugin incompatible with Kotlin Android experiemental extensions

### ISSUE THREAD

**Issue by btilbrook-nextfaze** [awslabs/aws-mobile-appsync-sdk-android#69]
**Describe the bug**
When using [Kotlin Android extensions experimental mode](https://kotlinlang.org/docs/tutorials/android-plugin.html#experimental-mode) in the app project, with AppSync used in a library project, the build fails.

**To Reproduce**
1. `cd` into attached project
1. `./gradlew clean assembleDebug`

**Sample project**
[AppSyncLibraryProject.zip](https://github.com/awslabs/aws-mobile-appsync-sdk-android/files/2622720/AppSyncLibraryProject.zip)

**Expected behavior**
Build succeeds.

**Actual behavior**
Build fails with compilation errors such as the following:

```
AppSyncLibraryProject/library/build/generated/source/appsync/type/CreateTodoInput.java:3: error: package com.apollographql.apollo.api does not exist
```

Commenting-out the following lines from `app/build.gradle` fixes it:

```
androidExtensions {
    experimental = true
}
```

**Environment:**
- AppSync SDK Version: `2.7.0`
- Kotlin `1.3.10`
- Android Gradle Plugin `3.3.0-rc01`
- Gradle `4.10.2` and `4.10.1`

**Additional context**
- AppSync is used in a library project, not the application project
- Project uses [Kotlin Android extensions experimental mode](https://kotlinlang.org/docs/tutorials/android-plugin.html#experimental-mode) in the application project

Also posted here, in case it's a Kotlin bug: https://youtrack.jetbrains.com/issue/KT-28548

**Comment by cbommas** [comment]
@btilbrook-nextfaze 
Thank you very much for providing the details on how to reproduce this issue. I was able to download the sample project and reproduce the error.

It appears that the problem is that java compiler is unable to pickup the Apollo imports in the generated code. I was able to get around it by adding the appsync SDK to main project build.gradle file

    dependencies {
        classpath 'com.android.tools.build:gradle:3.3.0-rc01'
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
        classpath "com.amazonaws:aws-android-sdk-appsync-gradle-plugin:2.7.0"
        classpath "com.amazonaws:aws-android-sdk-appsync:2.7.0"
    }

With this in place, both the assembleDebug and assembleRelease worked fine. Can you give this a shot?

**Comment by btilbrook-nextfaze** [comment]
Just tried making this change, but it didn't work for me. I tried it in the sample and my real project. I also tried adding it to `app/build.gradle`:

```
dependencies {
    // Other dependencies...
    implementation "com.amazonaws:aws-android-sdk-appsync:2.7.0" // also tried "api", same result
}
```

**Comment by cbommas** [comment]
@btilbrook-nextfaze 

I see that you are using `implementation` instead of `classpath`. If you use `classpath` in the buildscript dependencies section, it worked for me.  Here is the full top-level build.gradle file from your sample that I adjusted to make it run 

    // Top-level build file where you can add configuration options common to all sub-projects/modules.

    buildscript {
    ext.kotlin_version = '1.3.10'
    repositories {
        google()
        jcenter()

    }
    dependencies {
        classpath 'com.android.tools.build:gradle:3.3.0-rc01'
        classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
        classpath "com.amazonaws:aws-android-sdk-appsync-gradle-plugin:2.7.0"
        classpath  "com.amazonaws:aws-android-sdk-appsync:2.7.0"
       }
    }

     allprojects {
    repositories {
        google()
        jcenter()

      }
    }

    task clean(type: Delete) {
      delete rootProject.buildDir
     }

**Comment by btilbrook-nextfaze** [comment]
> I see that you are using implementation instead of classpath

Sorry, yes my example there was meant to indicate that I tried doing that in addition to your suggestion.

I did indeed also try adding the buildscript classpath line, without success.

**Comment by cbommas** [comment]
@btilbrook-nextfaze 
That is strange, as I can reproduce the error without the entry and then fix it by adding the entry.

It would be really helpful if you could attach the project one more time with the classpath entry added to the buildscript dependencies where you are encountering the error, so that I can reproduce it on my side.

**Comment by btilbrook-nextfaze** [comment]
Here you go:

[AppSyncLibraryProject 2.zip](https://github.com/awslabs/aws-mobile-appsync-sdk-android/files/2637453/AppSyncLibraryProject.2.zip)

**Comment by cbommas** [comment]
@btilbrook-nextfaze 

Thank you for uploading the new zip.

I was able to get this working by also adding `com.amazonaws:aws-android-sdk-appsync-api:2.7.0` to the top level and library level build files. 

Top level build.gradle
------------------
      buildscript {
          ext.kotlin_version = '1.3.10'
          repositories {
              google()
              jcenter()
           }
  
        dependencies {
             classpath 'com.android.tools.build:gradle:3.3.0-rc01'
             classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:$kotlin_version"
             classpath "com.amazonaws:aws-android-sdk-appsync-gradle-plugin:2.7.0"
             classpath "com.amazonaws:aws-android-sdk-appsync:2.7.0"
             classpath "com.amazonaws:aws-android-sdk-appsync-api:2.7.0"
         }
     }

    allprojects {
        repositories {
            google()
            jcenter()
       }
    }

     task clean(type: Delete) {
          delete rootProject.buildDir
     }

Library level build.gradle
-------------------------

    apply plugin: 'com.android.library' 
    apply plugin: 'kotlin-android'

    apply plugin: 'com.amazonaws.appsync'

    android {
        compileSdkVersion 28
        defaultConfig {
            minSdkVersion 23
        }
    }

    dependencies {
        implementation "com.amazonaws:aws-android-sdk-mobile-client:2.8.4"
        implementation "com.amazonaws:aws-android-sdk-core:2.8.4"
        implementation "com.amazonaws:aws-android-sdk-appsync:2.7.0"
        implementation "com.amazonaws:aws-android-sdk-appsync-api:2.7.0"
        implementation 'org.eclipse.paho:org.eclipse.paho.client.mqttv3:1.2.0'
        implementation 'org.eclipse.paho:org.eclipse.paho.android.service:1.1.1'
    }

Can you give this a try and let me know how you fare?

**Comment by btilbrook-nextfaze** [comment]
Those changes worked, thank you. I also had to add `aws-android-sdk-appsync-runtime`  to the library module in order to invoke AppSync Java APIs.

It seems the workaround is to add all of the transitive dependencies that would normally be added by the Gradle plugin and/or the main AppSync library dependency.

I have a feeling this is a Kotlin problem, since JetBrains seem to have [put it on their radar](https://youtrack.jetbrains.com/issue/KT-28548).

**Comment by cbommas** [comment]
@btilbrook-nextfaze 

Good to hear that it worked for you. I will close out the issue.

**Comment by btilbrook-nextfaze** [comment]
Appears to be fixed as of Kotlin `1.3.21`, AGP `3.3.1`.
